### PR TITLE
asset-swapper: Quicker path-finding

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

We can reduce the number of iterations we go through (`runLimit`) to find a decent path if we:
- Pre-sort all the individual (unmixed) paths by descending adjusted rate
- When mixing two paths, sort the fill nodes such that:
  - Fill nodes are always contiguous (form a valid subpath)
  - Subpaths (segments of the same source) are sorted by descending rate.    
- Compute the "rate" of a path based off `targetInput` (fill amount) instead of `min(targetInput, pathSize)`.
  - This favors paths that cover more of the total fill (not sure about this).
-  Increase the `runLimit` decay rate to `0.5` (prev `0.8`). So we halve the number of iterations for each subsequent call to `mixPaths()`.
- Start the `bestPath` in `mixPaths()` off with the left path (the previous mixed path), instead of starting with none.

This makes it more likely that we walk the viable paths first, so we can cut `runLimit` down to something like `2**8` (previously `2**13`).

This results in a much faster (2-3x) response time with small change in quotes:

ab-response-time

ab-realized

*Notes*: 
- WETH/DAI/USDC sims buy AND sell
- orange is this version w/ `runLimit = 2**8`
- blue is original version w/ `runLimit = 2**13`

We can probably tweak `runLimit` to improve results. This should buy us a little breathing room until we can entirely replace the algorithm with something better.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 1]
```diff
@@ -43,9 +44,10 @@ function mixPaths(
     targetInput: BigNumber,
     maxSteps: number,
 ): Fill[] {
-    let bestPath: Fill[] = [];
-    let bestPathInput = ZERO_AMOUNT;
-    let bestPathRate = ZERO_AMOUNT;
+    const _maxSteps = Math.max(maxSteps, 16);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:45]: Also cap the minimum so everyone gets *some* action.

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts [hunk 2]
```diff
@@ -285,11 +285,11 @@ export function getFallbackSourcePaths(optimalPath: Fill[], allPaths: Fill[][]):
...
 
 export function getPathAdjustedRate(side: MarketOperation, path: Fill[], targetInput: BigNumber): BigNumber {
-    const [input, output] = getPathAdjustedSize(path, targetInput);
-    if (input.eq(0) || output.eq(0)) {
+    const [, output] = getPathAdjustedSize(path, targetInput);
+    if (output.eq(0)) {
         return ZERO_AMOUNT;
     }
-    return side === MarketOperation.Sell ? output.div(input) : input.div(output);
+    return side === MarketOperation.Sell ? output.div(targetInput) : targetInput.div(output);
```
Comment (moodlezoup) [inline packages/asset-swapper/src/utils/market_operation_utils/fills.ts:273]: hm yeah idk about this –– have you AB tested this change alone?

File: packages/asset-swapper/src/utils/market_operation_utils/types.ts [hunk 3]
```diff
@@ -109,6 +109,10 @@ export enum FillFlags {
  * Represents a node on a fill path.
  */
 export interface Fill<TFillData extends FillData = FillData> {
+    // Unique ID of the original source path this fill belongs to.
+    // This is generated when the path is generated and is useful to distinguish
+    // paths that have the same `source` IDs but are distinct (e.g., Curves).
+    sourcePathId: string;
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/types.ts:126]: New field to tell consolidated sources apart.

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts [hunk 4]
```diff
@@ -112,6 +114,7 @@ function dexQuotesToPaths(
     ethToOutputRate: BigNumber,
     fees: FeeSchedule,
 ): Fill[][] {
+    const sourcePathId = hexUtils.random();
```
Comment (moodlezoup) [inline packages/asset-swapper/src/utils/market_operation_utils/fills.ts]: this should be inside the for loop I think

**Review by moodlezoup — COMMENTED** [review]
overall approach look good to me

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts (hunk 4 above)
Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/fills.ts]: ~I want it to be unique to the whole path. So we can tell which fill belongs to which curve.~

Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/fills.ts]: nvm ur rite

Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/fills.ts]: yay that fixed my sims

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 5]
```diff
@@ -57,29 +59,27 @@ function mixPaths(
...
+                if (!isValidPath(nextPath, true)) {
                     continue;
                 }
                 // Remove this fill from the next list of candidate fills.
                 const nextAllFills = allFills.slice();
                 nextAllFills.splice(i, 1);
                 // Recurse.
                 _walk(
-                    childPath,
+                    [...path, fill],
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts]:
```suggestion
                    nextPath,
```

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts (hunk 2 above)
Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/fills.ts:273]: `7001`: Using `targetInput` as `input` (this PR)
`7002`: Using `input` as `input` (original recipe)

ab-realized-rates

Comment (moodlezoup) [reply packages/asset-swapper/src/utils/market_operation_utils/fills.ts:273]: 🤷  whatever wins I guess

File: packages/asset-swapper/test/market_operation_utils_test.ts [hunk 6]
```diff
@@ -745,9 +739,9 @@ describe('MarketOperationUtils tests', () => {
                 expect(improvedOrders).to.be.length(3);
                 const orderFillSources = improvedOrders.map(o => o.fills.map(f => f.source));
                 expect(orderFillSources).to.deep.eq([
-                    [ERC20BridgeSource.Uniswap],
+                    [ERC20BridgeSource.Uniswap, ERC20BridgeSource.Curve],
                     [ERC20BridgeSource.Native],
-                    [ERC20BridgeSource.Eth2Dai, ERC20BridgeSource.Curve],
+                    [ERC20BridgeSource.Eth2Dai],
```
Comment (moodlezoup) [inline packages/asset-swapper/test/market_operation_utils_test.ts:818]: wait what's going on here?

Comment (dorothy-zbornak) [reply packages/asset-swapper/test/market_operation_utils_test.ts:818]: Yeah it's a little funky that Curve comes right after Uniswap now, but both ways are technically valid so I'm like w/e :shrug:

Comment (moodlezoup) [reply packages/asset-swapper/test/market_operation_utils_test.ts:818]: sus but ok

**Review by moodlezoup — APPROVED** [review]

File: packages/asset-swapper/test/market_operation_utils_test.ts (hunk 6 above)
Comment (dorothy-zbornak) [reply packages/asset-swapper/test/market_operation_utils_test.ts:818]: welcome to asset-swapper lol

**Pull request merged**